pgxn/neon/pagestore_fetchbuf.h

```diff
@@ -0,0 +1,77 @@
+#ifndef NEON_PAGESTORE_FETCHBUF_H
+#define NEON_PAGESTORE_FETCHBUF_H
+
+#include "storage/buf_internals.h"
+
+/*
+ * Prefetch implementation:
+ * 
+ * Prefetch is performed locally by each backend.
+ * 
+ * 
+ * There can be up to MAX_PREFETCH_REQUESTS registered using smgr_prefetch
+ * before smgr_read. All this requests are appended to primary smgr_read request.
+ * It is assumed that pages will be requested in prefetch order.
+ * Reading of prefetch responses is delayed until them are actually needed (smgr_read).
+ * It make it possible to parallelize processing and receiving of prefetched pages.
+ * In case of prefetch miss or any other SMGR request other than smgr_read,
+ * all prefetch responses has to be consumed.
+ */
+
+/* Max amount of tracked buffer reads */
+#define READ_BUFFER_SIZE 128
+
+typedef enum PrefetchStatus {
+	PRFS_UNUSED = 0, /* unused slot */
+	PRFS_REQUESTED, /* request is sent to PS, all fields except response valid */
+	PRFS_RECEIVED, /* all fields valid, response contains data */
+	PRFS_TAG_REMAINS, /* only buftag, *OfRel are still valid */
+} PrefetchStatus;
+
+typedef struct PrefetchRequest {
+	BufferTag	buftag; /* must be first entry in the struct */
+	XLogRecPtr	effective_request_lsn;
```

Comment:
This is never read. I think we should check this in `neon_read_at_lsn`, before we accept a prefetched response. It's possible that we issue a prefetch request with LSN 1000, and receive a response, but before we consume the response, another backend reads the same page, modifies it with LSN 2000, and it gets evicted again. In that case, we must not use the old prefetched page version at LSN 1000.

This is assigned from `prefetch_lsn`, in `prefetch_do_request_ext`, but `prefetch_lsn` is never set.